### PR TITLE
Krylov schur

### PR DESCRIPTION
**Pull request by michalk8**
Hi @msmdev 
I've added the returning of Schur matrix for `method='krylov'`, since @Marius1311 and I would also like to visualize it.

### CODE REVIEW

**Comment by Marius1311** [comment]
Hi @msmdev , can we merge this,  or would you like any changes?

**Comment by Marius1311** [comment]
Hi @msmdev , can we merge this,  or would you like any changes?

**Comment by Marius1311** [comment]
would be great if we could merge this over the weekend.

File: msmtools/util/sorted_schur.py [hunk 1]
```diff
@@ -352,7 +352,11 @@ def sorted_krylov_schur(P, m, z='LM'):
     # WE NEED REAL VECTORS! G-PCCA and PCCA only work with real vectors!!
     # We take the sequence of 1-D arrays and stack them as columns to make a single 2-D array.
     Subspace = np.column_stack([x.array for x in E.getInvariantSubspace()])
-    
+
+    R = E.getDS().getMat(SLEPc.DS.MatType.A)
```
Comment (msmdev) [inline msmtools/util/sorted_schur.py:356]: Have you tested that this really gives you the quasi-triangular matrix? (e.g. via comparison with R given back from brandts)

Comment (Marius1311) [reply msmtools/util/sorted_schur.py:356]: Mike has implemented this, I will test it and get back to you!

**Comment by Marius1311** [comment]
I will test this once we have merged #3, this will be easier as I need the lower convergence tolerance in krylov-schur to compare the returned matrices.

**Comment by Marius1311** [comment]
I looked into this a bit and wrote a utility function (see below) which checks for deviations between the schur forms returned by `brandts` and `krylov`, besides a couple of other comparison metrics like correlation between corresponding membership vectors. I iterated this for `m` between 2 and 20 states and report the results below. 

## Utility function
```python
def compare_brandts_krylov(T, n_states_max = 20):
    """Utility function to compare methods `krylov` and `brandts`
    

    Metrics for comparison are:
    - 'max_angle_svecs': maximum subspance angle between the first `n_states` schur vectors returned 
            by `brandts` and `krylov`
    - '1-min_corr_chi': 1 - the minimum corrleation between corresponding membership vectors computed by GPCCA 
            using the schur decomposition obtained using either `brandts` or `krylov`
    - 'max_diff_abs_sf': maximum element-wise difference between the absolute value in the schur forms returned
            by `brandts` and `krylov`
    - 'max_angle_sd_b': maximum subspace angle between the column spaces spanned by TX and XR, where X and R are the 
            schur vectors and schur form returned by `brandts` (restricted to the first `n_states`)
    - 'max_angle_sd_k': same as above, for `krylov`
    
    Params
    --------
    T
        Transition matrix
    n_states_max
        Maximum number of states. We iterate over `range(2, n_states_max)` and compute the above metrics for 
        all values in the range. 
        
    Returns
    --------
    restuls
        Pandas dataframe where `n_states` is the index and the above metrics form the columns
    """
    
    from msmtools.analysis.dense.gpcca import GPCCA
    from scipy.linalg import subspace_angles

    results = {}
    g_k = GPCCA(T, method='krylov')
    g_b = GPCCA(T, method='brandts')

    for n_states in range(2, n_states_max):

        # fit the estimators, skip if split pairs of compl. conj. 
        try:
            g_k.optimize(m=n_states)
            g_b.optimize(m=n_states)
        except ValueError:
            n_states += 1
            continue

        # get schur vectors, memberships and schur forms
        X_k, X_b = g_k.schur_vectors, g_b.schur_vectors
        chi_k, chi_b = g_k.memberships, g_b.memberships
        sf_k, sf_b = g_k.schur_matrix, g_b.schur_matrix

        # get max subspace angle and min correlation between membership vec's
        max_angle_schur_vecs = subspace_angles(X_k, X_b).max()
        max_angle_sd_b = np.max(subspace_angles(T.A @ X_b, X_b @ sf_b))
        max_angle_sd_k = np.max(subspace_angles(T.A @ X_k, X_k @ sf_k))
        min_corr_chi = np.min([np.corrcoef(a, b)[0, 1] for a, b in zip(chi_k.T, chi_b.T)])
        max_diff_abs_sf = np.max(np.abs(sf_k) - np.abs(sf_b))

        # save in dict
        results[n_states] = {'max_angle_svecs': max_angle_schur_vecs, 
                             '1 - min_corr_chi': 1 - min_corr_chi, 
                             'max_diff_abs_sf': max_diff_abs_sf,
                            'max_angle_sd_b': max_angle_sd_b,
                             'max_angle_sd_k': max_angle_sd_k,
                            }
        
    return pd.DataFrame.from_dict(results, orient='index')
```

## Comparison results


 n_states | max_angle_svecs | 1 - min_corr_chi | max_diff_abs_sf | max_angle_sd_b | max_angle_sd_k
-- | -- | -- | -- | -- | --
2 | 1.536182e-13 | 3.330669e-16 | 3.705297e-10 | 2.629617e-15 | 1.704016e-15
3 | 3.008713e-14 | 2.220446e-16 | 2.350319e-08 | 2.153157e-15 | 4.046935e-15
4 | 3.784091e-13 | 4.440892e-16 | 2.970367e-08 | 2.284389e-15 | 4.199154e-15
6 | 1.946549e-12 | 0.000000e+00 | 2.247817e-01 | 2.444560e-15 | 9.163733e-15
7 | 1.305682e-11 | 2.220446e-16 | 2.247817e-01 | 2.293919e-15 | 8.487898e-15
8 | 5.153335e-12 | 4.440892e-16 | 2.970367e-08 | 2.794926e-15 | 9.115425e-15
9 | 4.963620e-11 | 3.330669e-16 | 2.970367e-08 | 9.919000e-15 | 1.046712e-14
11 | 1.148596e-11 | 3.330669e-16 | 2.970367e-08 | 2.613812e-15 | 1.197667e-14
12 | 5.472191e-10 | 1.247890e-07 | 2.970367e-08 | 3.610327e-15 | 1.148576e-14
13 | 2.832554e-11 | 2.220446e-16 | 2.247817e-01 | 3.544062e-15 | 3.835541e-14
14 | 3.985297e-10 | 6.661338e-16 | 2.970367e-08 | 5.600004e-15 | 3.789700e-13
16 | 5.021773e-10 | 3.330669e-16 | 2.264099e-01 | 2.937583e-15 | 4.557657e-14
18 | 1.416328e-09 | 2.638015e-04 | 2.460048e-01 | 3.337323e-15 | 1.325748e-13
19 | 2.321983e-09 | 3.330669e-16 | 2.646620e-01 | 3.466842e-15 | 1.071683e-13



## Interpretation
Please see the doc string of the utility function for a definition of the metrics in the above table. In principle, this PR does indeed expose the krylov-schur computed real schur form matrix, which agrees element-wise in absolute value with the schur form computed by `brandts`, except for a number of values for `m` where there are maximum deviations of the order 1e-1. Except for the critical values `m=12` and `m=18`, the membership vectors agree very well.

@msmdev , what do you make of this? Any explanations why the schur forms sometimes disagree? The subspace angles between the column spaces spanned by the top schur vectors agree even in these cases. I'm not sure to what extent the schur form is unique.

**Pull request merged**